### PR TITLE
Separate span for input and output streams.

### PR DESCRIPTION
**Pull request by tylerbenson**
@realark I've separated the input and output streams as you requested.

I also pushed the `Tracer.inject` point a bit deeper to `HttpClient.writeRequests` where we know that it's actually happening.  Let me know if you'd rather I leave that how you had it.

### CODE REVIEW

File: dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/HttpUrlConnectionInstrumentation.java [hunk 1]
```diff
@@ -82,22 +69,37 @@ public static Scope startSpan(
...
-        state.setHasDoneIO(true);
+
+      switch (methodName) {
+        case "connect":
+          if (connected) {
+            return null;
+          }
+          // We get here in cases where connect() is called first.
+          // We need to inject headers now because after connected=true no more headers can be
+          // added.
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/HttpUrlConnectionInstrumentation.java]: Am I reading the code correctly and header injection is no longer happening in this class? And in fact it happens deeper so this code really doesn't control they way headers are injected anymore. In fact as far as I can see all this `switch` is controlling is creation of multiple spans when connect/getInputSpan/getOutputSpan are called multiple times.

If above is correct you may want to rewrite this comment to better clarify what is happening here.

File: dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/HttpUrlConnectionInstrumentation.java [hunk 2]
```diff
@@ -187,16 +187,9 @@ public static HttpURLState get(final HttpURLConnection connection) {
       return state;
     }
 
-    public boolean hasDoneIO = false;
+    public boolean calledOutputStream = false;
+    public boolean calledInputStream = false;
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/HttpUrlConnectionInstrumentation.java]: Above comment says `// not thread-safe, but neither is HttpURLConnection`. I think this is only partially true: `sun.net.www.protocol.http.HttpURLConnection` seems to be thread-safe. That's probably not a huge issue in reality, but comment may be confusing.

File: dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/SunHttpClientInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+  public static class InjectAdvice {
+
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void inject(
+        @Advice.Argument(0) final MessageHeader header, @Advice.This final HttpClient client) {
+      final Tracer tracer = GlobalTracer.get();
+      final Span span = tracer.activeSpan();
+      // Use the active span, not the one created in SpanStartAdvice,
+      // since we aren't sure that span will complete.
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/SunHttpClientInstrumentation.java]: Could you please clarify what this comment means and what is `SpanStartAdvice` it is referring to?

Comment (realark) [reply dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/SunHttpClientInstrumentation.java]: Is the purpose of this instrumentation to add header injection in cases where `connect()` is called before `get*Stream()`?

File: dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/HttpUrlConnectionInstrumentation.java (hunk 1 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/HttpUrlConnectionInstrumentation.java]: Updated.  Is it better?

File: dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/HttpUrlConnectionInstrumentation.java (hunk 2 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/HttpUrlConnectionInstrumentation.java]: I made it threadsafe.  Though I'm not sure if it's a problem synchronizing on the connection.

File: dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/SunHttpClientInstrumentation.java (hunk 3 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/SunHttpClientInstrumentation.java]: That comment is from a previous attempt at instrumenting the sun `HttpClient` further.  I've removed it.  Sorry for the confusion.

**Comment by tylerbenson** [comment]
@realark 
> Do we only create connect spans on the actual connection (when the connection is not cached)?

yes, it will create a connect span if connect is called.  This is to ensure a span exists for propagation.

> Does getInputStream do header injection in cases where the connection is cached (keep-alive) and connect() is called before getInputStream()?

It shouldn't.  It only does header injection right before the headers are written out to the socket, so in your question, it would be when connect is called.

**Pull request merged**